services/src/main/java/org/keycloak/protocol/oidc/mappers/AcrProtocolMapper.java

```diff
@@ -109,6 +109,7 @@ protected String getAcr(AuthenticatedClientSessionModel clientSession) {
         }
 
         Map<String, Integer> acrLoaMap = AcrUtils.getAcrLoaMap(clientSession.getClient());
+        // Do we want to add the enforcement and produce an error here in the mapper as well?
         String acr = AcrUtils.mapLoaToAcr(loa, acrLoaMap, AcrUtils.getRequiredAcrValues(
```

Comment:
@sonOfRa I think that responsibility of the `AcrProtocolMapper` should not be to do any additional checks, but just map the claim from the clientSession into the token.

The validations should be done in the layers (like for example `AuthorizationEndpoint`as you pointed) to make sure that authentication doesn't even start if the sent level in the `claims` is not correct and too low.

In the most ideal case, the validations should be done rather in the `AuthorizationEndpointChecker` rather than in `AuthorizationEndpoint.updateAuthenticationSession()` to make sure that request fail-fast even before authenticationSession is created (or also it will fail-fast for the PAR request). But I don't know if you want to update this point in your PR as ACR checks are already done before your PR in the `AuthorizationEndpoint.updateAuthenticationSession()`. So feel free to keep as it is in your PR (Just please remove the comment in `AcrProtocolMapper`)

Reply:
I'm open to moving it to `AuthorizationEndpointChecker` but to me it looks like that at least for acr, the checks and the part that updates the authenticationSession is quite closely intertwined with the setting of the client notes (belongs in `updateAuthenticationSession()` and the actual checking of the ACR values (belongs in `AuthorizationEndpointChecker`). I have no particularly strong feelings here.

I'll remove the comment, it was just a marker for me where to add the in-line review comment.

Reply:
@sonOfRa Yes, I see your point. I may take a look at this at some point, but not yet sure as it kind of works in a way it is done today (even if not 100% ideal as we may possibly handle it even before auth-session creation to possibly save some data etc). Thanks for removing the comment!